### PR TITLE
[Serializer] Add FlattenExceptionNormalizer / Fix messenger used with the serializer component

### LINKED ISSUE #32719: [Messenger] Message retried in failure transport dropped on exception
**Symfony version(s) affected**: 4.3.2

**Description**  
Message retried in failure transport dropped on exception

**How to reproduce**  
I configured a failure transport on symfony messenger (like explain here: https://symfony.com/blog/new-in-symfony-4-3-messenger-failure-transport)
I throw an error on my handler, the message is retried 3 times, and then sent to the failure transport.

Then, I retry the failed message with `bin/console messenger:consume failed` or `bin/console messenger:failed:retry {id}`.
The message handling fail again and the message is dropped.

Is it the normal behaviour ? Or is it a bug ?

### PR DESCRIPTION
**Pull request by skalpa**
| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #32719 
| License       | MIT

This adds a new normalizer for `FlattenException` objects.

It is against 4.3 and marked as a bug fix, because the Serializer component's current inability to deserialize `FlattenException` objects breaks Messenger when used with the component (see #32719 for details).

The new FlattenException class of the ErrorRenderer component suffers from the same problem, so once merged it will probably have to be amended on 4.4 to support the ErrorRenderer as well.

### CODE REVIEW

**Comment by skalpa** [comment]
The failing test on 7.3 is due to the fact that Travis loads the deps from `4.4.x-dev` and can't find the new class there, so I guess that's to be expected.

**Comment by mleczakm** [comment]
What with `traceAsString` field? Is it properly serialized/deserialized?

**Comment by skalpa** [comment]
@mleczakm That's fixed. Nice catch.

**Comment by mleczakm** [comment]
This fix would also help in https://github.com/symfony/symfony/issues/33049

**Review by weaverryan — APPROVED** [review]
Thanks for the PR!

I think this is the correct approach. I originally chose to use `FlattenException` because the purpose of it is to be a "serializable" exception. This, of course, works with Messenger's PhpSerializer. I think it makes sense to make it also work with Symfony's serializer. An alternative approach would be for Messenger to store the exception information in a simpler way (without using `FlattenException`), but that feels like reinventing the wheel: this is the purpose of `FlattenException`.

**Review by maxhelias — APPROVED** [review]

**Comment by dunglas** [comment]
Wouldn't it better to serialize the error following/extending the formats defined by RFC 7807 or Hydra (as we do for validation errors: #22150)?
I feel like we're duplicating login introduced in the new ErrorRenderer component?

Maybe that a better approach would be to:
* convert the renderer as simple normalizers, in the serializer component (as we've done for error validations)
* use the Serializer component as a dependency of the ErrorHandler component
* remove the ErrorRenderer component (it's not too late, it hasn't be tagged yet)

There are several benefits:

* the use case with Messenger is fixed
* API Platform will automatically support the new ErrorHandler component (API Platform already has normalizers for `\Exception` and `\FlattenException` for most modern JSON error formats including RFC 7807, Hydra and JSON:API)
* it removes yet another component dealing with serialization

At least, the error renderers should implement Serializer's interfaces (and these interfaces should probably moved to `symfony/contracts`.

/cc @yceruto

**Comment by skalpa** [comment]
@dunglas I'm not sure I follow how you'd expect the same set of classes to satisfy the needs of both Messenger and ErrorRenderer as the use cases are totally different.

Also, this PR is about fixing bugs in 4.3, so anything that involves components that haven't been released yet probably won't apply. 

* In Messenger, objects have to be serialized then deserialized with no loss of information (this PR is actually more about fixing deserialization than serialization).
* For error reporting, serialization is a one-way street, potentially destructive, and context dependent (i.e: AFAIK the behaviour of both the api-platform normalizers and ErrorRenderer depend on the value of `debug`).

As for following RFC 2807 more closely, I'm all for it. However as some fields in the RFC only make sense in an HTTP context, there's not much we could do I guess. Would you have anything else than the following in mind ?

* `type`, `title` and `instance` will have to be omitted (that's ok if I understand the RFC)
* Rename `status_code` to `status`
* Rename `message` to `detail`

**Comment by yceruto** [comment]
@dunglas I failed to see as well the approach described and its relationship with this PR + ErrorRenderer component in 4.4.

The ErrorRenderer component does not worry (and shouldn't) about the full normalization of the FlattenException object nor its denormalization at all, but standardized error responses in different formats, included the HTML error/exception page.

I don't know how all those features would fit into the Serializer component instead.

**Comment by dunglas** [comment]
@ skalpa

> the needs of both Messenger and ErrorRenderer as the use cases are totally different

They don't look that different to me. Also, the benefit of using JSON with Messenger (instead of the native PHP `serialize()` format) is to be able to consume the messages with other programming languages. It's exactly there are standards for error representations. And converting an internal data structure (such as `\Exception`) to a public (or even better, standard) representation that can be consumed by another system is exactly the responsibility of the Serialize component.

> In Messenger, objects have to be serialized then deserialized with no loss of information

It can be added to extra keys, as done by API Platform (such information could also be useful for developers trying to debug). See https://github.com/api-platform/core/blob/master/src/Hydra/Serializer/ErrorNormalizer.php#L58-L60 for instance.

> For error reporting, serialization is a one-way street

Normalizers in API Platform currently only support normalization, but there are no technical reasons preventing to support deserialization.

> potentially destructive, and context dependent (i.e: AFAIK the behaviour of both the api-platform normalizers and ErrorRenderer depend on the value of debug).

We have all the infrastructure we need in the Serializer to deal with this. An option could be passed to the context (and we can now set default context options easily) to choose if debug informations for instance should be serialized or not.

> As for following RFC 7807 more closely, I'm all for it. However as some fields in the RFC only make sense in an HTTP context, there's not much we could do I guess. Would you have anything else than the following in mind ?

Great. I indeed had something like that in mind.

@yceruto 

> but standardized error responses in different formats

For instance, in API Platform we already support generating standard error response for RFC 7807, JSON:API and Hydra using normalizers. 

* https://github.com/api-platform/core/blob/master/src/Problem/Serializer/ErrorNormalizer.php
* https://github.com/api-platform/core/blob/master/src/Hydra/Serializer/ErrorNormalizer.php
* https://github.com/api-platform/core/blob/master/src/JsonApi/Serializer/ErrorNormalizer.php

If the ErrorHandler component could use any class implementing the `SerializerInterface` or (the `NormalizerInterface`), it would allow to reuse those normalizers very easily. Also, FOSRest comes with similar normalizers, and I'm such we can find others. To me it's the responsibility of the Serializer to transform a PHP object (`\Exception` here) in a standard open format (RFC 7807...).

> included the HTML error/exception page. It's more like (https://github.com/filp/whoops)

For (advanced) HTML like this agree. Creating something like whoops is totally out of scope of the Symfony Serializer and should stay in ErrorRenderer. However, it's not the same case for non-HTML renderers. They should be moved to the Serializer component.
Even if it doesn't belong to the Serializer component, the HTML component could also easily implement the `SerializerInterface` (it still convert an \Exception to a string).

**Comment by skalpa** [comment]
@dunglas I renamed `status_code` and `message` as discussed. As the RFC states that `status` should be a number if present, it is not in the normalized array if the exception has no status code set.

Regarding your other points, I now understand your point of view better, thank you. However, I believe this is a different discussion and don't think this should keep us from merging this PR, which is only for the `FlattenException` class of the Debug component at this point. 

We'll still be able to add support for the ErrorRenderer class and/or use the Serializer in the new components on 4.4 once this is done.

**Comment by fabpot** [comment]
ping @dunglas

**Comment by dunglas** [comment]
We can merge this one, but can we mark it as @experimental in case we have to modify it for ErrorHandler/ErrorRenderer?

**Comment by skalpa** [comment]
@dunglas Depending on what you decide to do, it might be best to create another normalizer for ErrorRenderer, and just retire this one with the Debug component in 5.0.

Anyway, I added the `@experimental` tag as it can't hurt.

**Comment by nicolas-grekas** [comment]
Does #34312 change the approach proposed here? Can you please have a look for 4.4?

**Comment by monteiro** [comment]
I suppose the only thing needed to merge this PR is to change this PR to use the new error handler component (and remove symfony/debug old component) right and put this agaisn't 4.4?

**Comment by monteiro** [comment]
I found the problem with the error. For some reason the Flatten Normalizer in this PR was not being called when normalizing the `FlattenException` object.

If I omit the priority on the Flatten Normalizer (in this PR) everything works as expected and the command `bin/console messenger:failed:retry -vv` does not fail.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/serializer.xml [hunk 1]
```diff
@@ -31,6 +31,11 @@
         <service id="Symfony\Component\Serializer\Mapping\ClassDiscriminatorResolverInterface" alias="serializer.mapping.class_discriminator_resolver" />
 
         <!-- Normalizer -->
+        <service id="serializer.normalizer.flatten_exception" class="Symfony\Component\Serializer\Normalizer\FlattenExceptionNormalizer">
+            <!-- Run before serializer.normalizer.object -->
+            <tag name="serializer.normalizer" priority="-915" />
```
Comment (monteiro) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/serializer.xml:36]: If I remove the priority the normalize() method is called. 
denormalize is called with our without the priority.

With this PR (the custom normalizer of this PR was not being called, that's why the trace_as_string is not being set). Result in xdebug:
Captura de ecrã 2020-05-28, às 08 47 24

(just did a throw new \InvalidArgumentException() in an handler to force failing)

File: src/Symfony/Component/Serializer/composer.json [hunk 2]
```diff
@@ -25,6 +25,7 @@
         "symfony/property-access": "~3.4|~4.0",
         "symfony/http-foundation": "~3.4|~4.0",
         "symfony/cache": "~3.4|~4.0",
+        "symfony/debug": "~3.4|~4.0",
```
Comment (monteiro) [inline src/Symfony/Component/Serializer/composer.json:28]: error-handler now?

**Comment by nicolas-grekas** [comment]
Continued in #37087, thank you @skalpa

**Pull request closed without merging**